azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -539,80 +853,97 @@ private void applyDisabledJobs(List<Object> disabledJobs,
+            message =
               "Flow " + flowId + " is already running with exec id "
-                  + StringUtils.join(running, ",")
-                  + ". Will execute concurrently. \n";
+                + StringUtils.join(running, ",")
+                + ". Will execute concurrently. \n";
+          }
         }
-      }
 
-      boolean memoryCheck = !ProjectWhitelist.isProjectWhitelisted(exflow.getProjectId(),
-              ProjectWhitelist.WhitelistType.MemoryCheck);
-      options.setMemoryCheck(memoryCheck);
-
-      // The exflow id is set by the loader. So it's unavailable until after
-      // this call.
-      executorLoader.uploadExecutableFlow(exflow);
-
-      // We create an active flow reference in the datastore. If the upload
-      // fails, we remove the reference.
-      ExecutionReference reference =
-          new ExecutionReference(exflow.getExecutionId(), executorHost,
-              executorPort);
-      executorLoader.addActiveExecutableReference(reference);
-      try {
-        callExecutorServer(reference, ConnectorParams.EXECUTE_ACTION);
-        runningFlows.put(exflow.getExecutionId(),
-            new Pair<ExecutionReference, ExecutableFlow>(reference, exflow));
+        boolean memoryCheck =
+          !ProjectWhitelist.isProjectWhitelisted(exflow.getProjectId(),
+            ProjectWhitelist.WhitelistType.MemoryCheck);
+        options.setMemoryCheck(memoryCheck);
+
+        // The exflow id is set by the loader. So it's unavailable until after
+        // this call.
+        executorLoader.uploadExecutableFlow(exflow);
+
+        // We create an active flow reference in the datastore. If the upload
+        // fails, we remove the reference.
+        ExecutionReference reference =
+          new ExecutionReference(exflow.getExecutionId());
 
+        if (isMultiExecutorMode()) {
+          //Take MultiExecutor route
+          executorLoader.addActiveExecutableReference(reference);
+          queuedFlows.enqueue(exflow, reference);
+        } else {
+          // assign only local executor we have
+          reference.setExecutor(activeExecutors.iterator().next());
+          executorLoader.addActiveExecutableReference(reference);
+          try {
+            callExecutorServer(reference, ConnectorParams.EXECUTE_ACTION);
+            runningFlows.put(exflow.getExecutionId(),
+              new Pair<ExecutionReference, ExecutableFlow>(reference, exflow));
+          } catch (ExecutorManagerException e) {
+            executorLoader.removeActiveExecutableReference(reference
+              .getExecId());
+            throw e;
+          }
+        }
         message +=
-            "Execution submitted successfully with exec id "
-                + exflow.getExecutionId();
-      } catch (ExecutorManagerException e) {
-        executorLoader.removeActiveExecutableReference(reference.getExecId());
-        throw e;
+          "Execution submitted successfully with exec id "
```

Comment:
it is outside the catch but inside the synchronized block. Catch completes at line 942